File: occasum-core/src/main/java/pw/stamina/occasum/node/factory/inspect/NameValidatingPropertyNodeFactoryParameterInspector.java
package pw.stamina.occasum.node.factory.inspect;

import pw.stamina.occasum.Named;
import pw.stamina.occasum.PropertyHandle;
import pw.stamina.occasum.node.PropertyNode;
import pw.stamina.occasum.properties.Property;

final class NameValidatingPropertyNodeFactoryParameterInspector
        implements PropertyNodeFactoryParameterInspector {

    @Override
    public void inspectRoot(PropertyHandle handle) {
        validatePropertyNodeName(handle.getName());
    }

    @Override
    public void inspectProperty(PropertyHandle handle, PropertyNode parent, Property property) {
        validatePropertyNodeName(property.getName());
    }

    @Override
    public void inspectFolder(PropertyHandle handle, PropertyNode parent, String name) {
        validatePropertyNodeName(name);
    }

    private static void validatePropertyNodeName(String name) throws IllegalArgumentException {
        Named.validateName(name);

        if (name.equals("value")) {
            throw new IllegalArgumentException(
                    "the name \"value\" is reserved for serialization purposes");
        }
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/node/factory/inspect/NameValidatingPropertyNodeFactoryParameterInspector.kt
package pw.stamina.occasum.node.factory.inspect

import pw.stamina.occasum.Named
import pw.stamina.occasum.PropertyHandle
import pw.stamina.occasum.dao.PropertyDao
import pw.stamina.occasum.node.PropertyNode
import pw.stamina.occasum.properties.Property

internal class NameValidatingPropertyNodeFactoryParameterInspector : PropertyNodeFactoryParameterInspector {

    override fun inspectRoot(handle: PropertyHandle) {
        validatePropertyNodeName(handle.name)
    }

    override fun inspectProperty(handle: PropertyHandle, parent: PropertyNode, property: Property<*>) {
        validatePropertyNodeName(property.name)
    }

    override fun inspectFolder(handle: PropertyHandle, parent: PropertyNode, name: String) {
        validatePropertyNodeName(name)
    }

    @Throws(IllegalArgumentException::class)
    private fun validatePropertyNodeName(name: String) {
        Named.validateName(name)

        if (PropertyDao.RESERVED_SERIALIZED_VALUE_NAME == name) {
            throw IllegalArgumentException("the name 'value' is reserved for serialization purposes")
        }
    }
}
